From 283a0a7ce8b42e7ea62b3b82f514a2564a99eae5 Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Thu, 20 Jul 2023 14:00:53 +0800
Subject: [PATCH 10/53] [Huawei] mm/sharepool: Ignore the mmap failure in
 add_group

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

There is a mechanism change that we don't break and return an error when
the mmap for the existing spa failed. We ignore this error for spa of
type SPA_TYPE_K2U already.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 49 +++++--------------------------------------------
 1 file changed, 5 insertions(+), 44 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index d22cf32857dd..c5a712d73ff9 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -1203,38 +1203,6 @@ static struct sp_group *find_or_alloc_sp_group(int spg_id)
 
 static void __sp_area_drop_locked(struct sp_area *spa);
 
-/* The caller must down_write(&mm->mmap_lock) */
-static void sp_munmap_task_areas(struct mm_struct *mm, struct sp_group *spg, struct list_head *stop)
-{
-	struct sp_area *spa, *prev = NULL;
-	int err;
-
-
-	spin_lock(&sp_area_lock);
-	list_for_each_entry(spa, &spg->spa_list, link) {
-		if (&spa->link == stop)
-			break;
-
-		__sp_area_drop_locked(prev);
-		prev = spa;
-
-		atomic_inc(&spa->use_count);
-		spin_unlock(&sp_area_lock);
-
-		err = do_munmap(mm, spa->va_start, spa_size(spa), NULL);
-		if (err) {
-			/* we are not supposed to fail */
-			pr_err("failed to unmap VA %pK when munmap task areas\n",
-			       (void *)spa->va_start);
-		}
-
-		spin_lock(&sp_area_lock);
-	}
-	__sp_area_drop_locked(prev);
-
-	spin_unlock(&sp_area_lock);
-}
-
 /* the caller must hold sp_group_sem */
 static int mm_add_group_init(struct task_struct *tsk, struct mm_struct *mm,
 			     struct sp_group *spg)
@@ -1521,22 +1489,18 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 
 		down_write(&mm->mmap_lock);
 		if (unlikely(mm->core_state)) {
-			sp_munmap_task_areas(mm, spg, &spa->link);
 			up_write(&mm->mmap_lock);
-			ret = -EBUSY;
-			pr_err("add group: encountered coredump, abort\n");
+			pr_warn("add group: encountered coredump\n");
 			spin_lock(&sp_area_lock);
-			break;
+			continue;
 		}
 
 		addr = sp_mmap(mm, file, spa, &populate, prot_spa, NULL);
 		if (IS_ERR_VALUE(addr)) {
-			sp_munmap_task_areas(mm, spg, &spa->link);
 			up_write(&mm->mmap_lock);
-			ret = addr;
-			pr_err("add group: sp mmap failed %d\n", ret);
+			pr_warn("add group: sp mmap failed %ld\n", (long)addr);
 			spin_lock(&sp_area_lock);
-			break;
+			continue;
 		}
 		up_write(&mm->mmap_lock);
 
@@ -1547,11 +1511,8 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 			else
 				pr_warn_ratelimited("add group failed, mm populate failed (potential no enough memory when -12): %d, spa type is %d\n",
 						ret, spa->type);
-			down_write(&mm->mmap_lock);
-			sp_munmap_task_areas(mm, spg, spa->link.next);
-			up_write(&mm->mmap_lock);
 			spin_lock(&sp_area_lock);
-			break;
+			continue;
 		}
 
 		spin_lock(&sp_area_lock);
-- 
2.17.1

